### ISSUE TITLE
samtools view with -L, -M and --fetch-pairs doing full CRAM traversal

### ISSUE THREAD

**Issue by acorvelo** [samtools/samtools#2059]
samtools 1.20
htslib 1.20
Linux 3.10.0-1160.71.1.el7.x86_64

Hi,
In samtools view, when using a BED file to extract overlapping reads and their mates, using the options `-M -L intervals.bed --fetch-pairs`, samtools appears to traverse the entire CRAM on the second (I assume) pass, leading to huge run times, even for a small number of intervals. The same doesn't appear to be the case if the interval(s) is/are specified at the end of the command using `chr:start-end` notation. Is this difference in behavior intended? Or, is there another way to invoke it with the `-L` option that avoids full traversals?

Thanks!

**Comment by daviesrob** [comment]
It's difficult to say without seeing the input files, but it's likely that the behaviour is coming from the `--fetch-pairs` option.  If the paired reads are spread widely over the genome, it may have to read a large proportion of the input cram on the second pass.  It will try to skip over areas that are not needed if it can.

This is more likely to happen with CRAM files compared to BAM, as CRAM stores data in blocks of 10000 alignment records by default.  If a single pair hits a block, the entire block will need to be decoded to get its record.  You can change this by using an [option](https://www.htslib.org/doc/samtools.html#GLOBAL_COMMAND_OPTIONS) like `--output-fmt-option seqs_per_slice=1000` when writing your cram file.  The resulting file will be bigger but more efficient for random access.

I'm not sure why putting the locations to search on the command-line made a difference compared to them being in a BED file, unless the list on the command-line test was shorter than the other one?

**Comment by acorvelo** [comment]
@daviesrob,
after further investigation, I can confirm that specifying the region at the end of the command or in a BED file makes no difference. My mistake there. I think I was using different intervals, hence the difference in behavior. I'm also now convinced that `samtools` is not doing a full traversal as that would take longer than the run times I'm observing.

That said, there is still an issue and it doesn't have to do with the number of intervals or the number of blocks that need to be decoded to extract the mates.

Let's first consider two similar single-end extractions (the first interval starts 1 base upstream of the start of the second interval):
```
$ time samtools view -@ 8 -T ${REFERENCE} -F 3840 ${CRAM} chr1:69861-69863 > 1.sam ;
0.14user 0.12system 0:00.28elapsed 94%CPU (0avgtext+0avgdata 276844maxresident)k
0inputs+0outputs (0major+9091minor)pagefaults 0swaps

$ time samtools view -@ 8 -T ${REFERENCE} -F 3840 ${CRAM} chr1:69862-69863 > 2.sam ;
0.12user 0.14system 0:00.27elapsed 97%CPU (0avgtext+0avgdata 276848maxresident)k
0inputs+0outputs (0major+9102minor)pagefaults 0swaps

$ wc -l 1.sam 2.sam 
   73 1.sam
   73 2.sam
  146 total

$ md5sum 1.sam 2.sam ; 
8114c3b32125fefb93773961afbf80e5  1.sam
8114c3b32125fefb93773961afbf80e5  2.sam
``` 
Both extractions  yield exactly the same 73 reads and take approximately the same time (< 1sec)

Now, fetching the mates as well:
```
$ time samtools view -@ 8 -T ${REFERENCE} -F 3840 --fetch-pairs ${CRAM} chr1:69861-69863 > 3.sam ;
0.17user 0.18system 0:00.55elapsed 65%CPU (0avgtext+0avgdata 277280maxresident)k
0inputs+0outputs (0major+9425minor)pagefaults 0swaps

$ time samtools view -@ 8 -T ${REFERENCE} -F 3840 --fetch-pairs ${CRAM} chr1:69862-69863 > 4.sam ;
938.30user 22.73system 2:35.51elapsed 617%CPU (0avgtext+0avgdata 306004maxresident)k
0inputs+0outputs (0major+4409101minor)pagefaults 0swaps

$ wc -l 3.sam 4.sam 
   138 3.sam
   138 4.sam
   276 total

$ md5sum 3.sam 4.sam ; 
f07211fb905b3f3acdcd2eb4ab3f7da3  3.sam
f07211fb905b3f3acdcd2eb4ab3f7da3  4.sam

``` 

The second `samtools` command (`chr1:69862-69863`), despite the fact that the interval specified was actually smaller (though with the same set of overlapping reads), took more than 2 mins to run, yielding the same result as the first one. I would assume that the second pass through the CRAM should be identical to that of the first command (`chr1:69861-69863`) but it's taking significantly longer overall (>2.5mins vs. < 1 sec) and with high CPU usage throughout.

I consistently get very similar run times for different iterations of this particular exercise. I'm also observing similar behavior for other intervals on the same CRAM and also on different CRAMS. Unfortunately, those are not consented for sharing and I didn't manage yet to reproduce the behavior on publicly available files, such as those from the 1000 Genomes Project, for instance.

**Comment by jkbonfield** [comment]
I don't think it's likely page faults.  It's taking ~5000x more user-time CPU and ~100x more system CPU, but has only ~50x as many page faults, so the rate of fault is actually down.

More telling is the huge CPU.  To me that implies something algorithmic.  I agree though that a perf record/report may help, albeit without the `-e` to profile where the CPU is being spent in both scenarios.  It may just be the same area simply scaled up, which implies a potential index related issue and simply decoding (and discarding) more data.

**Comment by jkbonfield** [comment]
There's definitely something terribly broken with the `--fetch-pairs` code.  

```
$ io_trace -S -r 'cram$' -- samtools view NA12878_S1.cram chr1:69862-69863 

File: NA12878_S1.cram
    Num. open    	1
    Num. close   	1
    Num. seeks   	1
    Num. read    	8
    Bytes read   	488114

$ io_trace -S -r 'cram$' -- samtools view --fetch-pairs NA12878_S1.cram chr1:69862-69863 
File: NA12878_S1.cram
    Num. open    	1
    Num. close   	1
    Num. seeks   	3
    Num. read    	31366
    Bytes read   	3919197460
```

So it's done a huge amount of data fetching and consequential decoding, with a big CPU cost.  Why?
Let's look at the spread of those pairs, based on the RNEXT and PNEXT we can see there are no big jumps.  Every read is paired nearby.

```
samtools view NA12878_S1.cram chr1:69862-69863 | awk '{print $7,$8}'|sort -k 2n
= 69591
= 69611
= 69611
= 69613
...
= 70097
= 70097
= 70112
```

So if we fetch all the records for chr1:69591-70112 then we should do precisely what it states in the man page:  do a range query, identify PNEXT/RNEXT, produce new ranges and then do a second query to fetch the pairs, and then filter by name again.  Only this is much MUCH less I/O:

```
$ io_trace -S -r 'cram$' -- samtools view NA12878_S1.cram chr1:69591-70112 

File: NA12878_S1.cram
    Num. open    	1
    Num. close   	1
    Num. seeks   	1
    Num. read    	8
    Bytes read   	488114
```

Infact it's identical I/O to the original query as both ranges fall within the same CRAM container.

This is proof that fetch-pairs is not working as documented and not as intended.  It's curious that it didn't do a large number of seeks, so it's not repeatedly refetching the same data.  I think it's just reading to the end of the chromosome or something odd.

**Comment by jkbonfield** [comment]
Ah, it's gone to chromosome `*` as `iter->nocoor` is true.

This is due to incorrect logic in the fetch-pairs code.  Specifically https://github.com/samtools/samtools/blob/develop/sam_view.c#L642-L658 has:

```
if ( rec->core.mtid < 0 || (rec->core.flag & BAM_FMUNMAP) ) nunmap = 1;
```

My little test data had this in it:

```
samtools view NA12878_S1.cram -M chr1:69862-69863 chr1:69591-70112 | grep H
SQ1004:134:C0D8DACXX:3:1205:9020:154773
HSQ1004:134:C0D8DACXX:3:1205:9020:154773	73	chr1	69790	0	101M	=	69790	0ACTGCTCACATTACAGTAGTTCTTTTGTTCTTTGGACCATGTGTCTTTATTTATGCCTGGCCATTCCCCATCAAGTCATTAGATAAATTCCTTGCTGTATT	CCCFFFFFHHHHHJJJIJJJIHIJJJJJJJJJJJJIIJJJJIJJJIIJHIJJJJJIIJJJJJIIIJHHFHEHIGIGEGHHGHGHHHHHC@CFFFCEDCEEE	XT:A:R	SM:i:0	AM:i:0	X0:i:3	X1:i:0	XM:i:0	XO:i:0	XG:i:0	XA:Z:chr15,-102462463,101M,0;chr19,+111378,101M,0;	MD:Z:101	NM:i:0	RG:Z:NA12878
HSQ1004:134:C0D8DACXX:3:1205:9020:154773	133	chr1	69790	0	*	=	69790	0TATTTAAGAAATAAAAAAAAAAAAATTTTAAAAAAAAAAAAATATATTTTTTTTTTTTAAAAAAATAAAAAAAAATTATTAAAAAAAAAAATAAAATAAAA	+114+,,+++2<+++<??GE<GE0'(.)))))))-,',850&((+(+((++)&000&&((((()&(((+(+&&05((+++(+((()&)&)&((((((((((	XC:i:35	RG:Z:NA12878
```

So two reads with flags 73 and 133, corresponding to

```
0x49	73	PAIRED,MUNMAP,READ1
0x85	133	PAIRED,UNMAP,READ2
```

So yes, the mate is unmapped.  However it is *placed*.  Adding chromosome `HTS_IDX_NOCOOR` into the regions to search for is incorrect!  The mate ref-name and mate ref-pos are the correct fields to use for looking for the data and not `BAM_FMUNMAP`.

That said, it's quite possible for some regions we may find data that is in the unmapped region of the file and that will always inevitably be tragically slow to query as it's unsorted.  Although ideally it shouldn't happen as we usually put unmapped reads adjacent to their mapped reads (but note they are size 1, so we can get the mapped data without its mapped sibling, which is why reducing the query size leads to this issue).

This patch cures the slow down I'm seeing, although it's oddly still doing more seeks and more reads than I'd expect.  Maybe that's just because it does two queries and doesn't exclude the first one?  Not sure.  Good enough though.

```
diff --git a/sam_view.c b/sam_view.c
index 7ec34a00..3125702e 100644
--- a/sam_view.c
+++ b/sam_view.c
@@ -639,7 +639,7 @@ static int fetch_pairs_collect_mates(samview_settings_t *conf, hts_itr_multi_t *
             }
         }
 
-        if ( rec->core.mtid < 0 || (rec->core.flag & BAM_FMUNMAP) ) nunmap = 1;
+        if ( rec->core.mtid < 0 ) nunmap = 1;
         if ( rec->core.mtid >= 0 ) {
             if (_reglist_push(&conf->reglist, &conf->nreglist, rec->core.mtid, rec->core.mpos,rec->core.mpos+1) != 0)
                 goto out;
```

@acorvelo if you patch samtools/sam_view.c with this does it cure the problem for you?

**Comment by acorvelo** [comment]
@jkbonfield , thank you for looking into this. I was out and unable to test the patch but what you describe perfectly matches the fact that the CRAMs where I'm observing the behavior are characterized by a large number of fully unmapped read pairs, hence the large run times. Also, problematic intervals always overlap one or more unmapped reads. I had previously coded a wrapper that does pretty much what is described in samtools view manual for `--fetch-pairs`, that only used `QNAME`, `RNEXT` and `PNEXT` and that worked as intended (if not for the meanwhile addressed #2063) so [972c188](https://github.com/samtools/samtools/commit/972c1889942a4f07d8f62e93330f723da919c271) should hopefully solve it.

Thanks again for quickly addressing this and looking forward for the next release. Any expectation of when that's coming out?

**Comment by jkbonfield** [comment]
New releases tend to be every 3-4 months, but that can be longer if we're busy with other projects and not much has changed in samtools et al.  So I wouldn't expect one until maybe August.

Meanwhile this particular problem could be fixed by retrofitting a one line change (removal of `|| (rec->core.flag & BAM_FMUNMAP)`) to 1.20 and using a local build.
